src/main/java/org/folio/util/WebClientFactory.java

```diff
@@ -0,0 +1,31 @@
+package org.folio.util;
+
+import static org.folio.rest.RestVerticle.MODULE_SPECIFIC_ARGS;
+import static org.folio.util.Constants.DEFAULT_TIMEOUT;
+import static org.folio.util.Constants.LOOKUP_TIMEOUT;
+
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.client.WebClientOptions;
+
+public class WebClientFactory {
+
+  private WebClient client;
```

Comment:
This client runs on a single Vert.x Context.
It should run on the Vert.x Context of the caller.
Each Verticle has a different Vert.x Context.
When the mod-login Verticle is restarted or two mod-login Verticles are needed the WebClient should be on the correct Vert.x Context.
There should be exactly one WebClient per Vert.x Context = per Verticle. This ensures that
* all task of a Verticle run on one thread so that we don't need to care about parallelism
* all connections of one Verticle go through one WebClient to allow for pooling, keep alive, and pipe-lining: https://vertx.io/docs/vertx-core/java/#_http_1_x_pooling_and_keep_alive

Reply:
This unit tests shows that the Vert.x Context and the Thread are wrong.
```
package org.folio.util;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Context;
import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(VertxUnitRunner.class)
public class WebClientFactoryTest {

  public class MyVerticle extends AbstractVerticle {
    @Override
    public void start(Future<Void> startFuture) {
      Thread verticleThread = Thread.currentThread();
      WebClientFactory.getWebClient().getAbs("http://invalid/").send(get -> {
        Context clientContext = Vertx.currentContext();
        if (clientContext != context) {
          startFuture.fail("Context mismatch:\n" + context + " !=\n" + clientContext
              + "\n Thread mismatch:\n" + verticleThread + " != \n" + Thread.currentThread());
        }
      });
    }
  }

  @Test
  public void verticles(TestContext testContext) {
    Vertx vertx = Vertx.vertx();
    Vertx.vertx().deployVerticle(new MyVerticle(), testContext.asyncAssertSuccess(verticle1 -> {
      vertx.deployVerticle(new MyVerticle(), testContext.asyncAssertSuccess());
    }));
  }
}
```

Reply:
OK I'll take a look later this morning

Reply:
I'm pretty sure this means a singleton based on a static member isn't going to work.  Where does that leave us?  Back at storing the client on the context itself, or having the WebClientFactory store clients on a per-context basis, then do a lookup and return the appropriate client for the provided context.

Reply:
How likely is it that we're ever going to run multiple mod-login verticles in the same JVM?  Wouldn't that typically be done by running multiple containers or at least java processes?

Reply:
The current WebClientFactory implementation executes all async callback handlers on a different thread than the mod-login methods called by RMB RestVerticle. This may cause concurrency problems that can be avoided by using one WebClient for one Vert.x Context.

Reply:
Tim Fox [says](https://groups.google.com/d/msg/vertx/g6uM5QP5HBo/35i4tV8aAgAJ):
> you're just calling some code in a verticle class from a main thread. You shouldn't do this with any verticle as it will break the Vert.x threading model. Verticles should only be called from a thread corresponding to their context.

Blake [says](https://groups.google.com/d/msg/vertx/oy0XvkkkJAk/crsdwyI5AAAJ):
> I'm not sure if you've got multiple verticles going and therefore multiple contexts and eventloops, but if you do and you're sharing a webclient (and thus an HttpClient) between them, then they're going to run into issues. You should create at least one client per verticle rather than one for multiple verticles.

Jason Porter [says](https://groups.google.com/d/msg/vertx/qgrUTMOObr0/juT2xtb9CgAJ):
> Turns out we had a static WebClient that was trying to be reused between runs.